### PR TITLE
Fix messaging [WIP]

### PR DESCRIPTION
**Pull request by sgdc3**
Messages are not properly forwarded to the bungeecord network.

TODO:
- [x] make login and logout broadcast messages (involves updating AuthMeBungee probably)
- [x] move the (database instanceof CacheDataSource) check login into a proper service (maybe? feedback pls)

### CODE REVIEW

File: src/main/java/fr/xephi/authme/service/bungeecord/BungeeSender.java [hunk 1]
```diff
@@ -75,13 +75,17 @@ public void connectPlayerOnLogin(Player player) {
...
-            if(!plugin.isEnabled()) {
+            if (!plugin.isEnabled()) {
                 ConsoleLogger.debug("Tried to send a " + type + " bungeecord message but the plugin was disabled!");
                 return;
             }
-            sendBungeecordMessage("AuthMe", type, playerName.toLowerCase());
+            if (type.isBroadcast()) {
+                sendBungeecordMessage("Forward", "ALL", "AuthMe", type.getId(), playerName.toLowerCase());
+            } else {
+                sendBungeecordMessage("AuthMe", type.getId(), playerName.toLowerCase());
```
Comment (games647) [inline src/main/java/fr/xephi/authme/service/bungeecord/BungeeSender.java]: This won't work in [1.13. Channels have to be in the following format:](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/browse/src/main/java/org/bukkit/NamespacedKey.java#32)
Namespace, Key := [a-z0-9._-]
Channel := Namespace:Key

and for 1.7.X support [smaller than 16 characters in total](https://github.com/bergerkiller/CraftSource/blob/8e1126adbbde1096b8351650bdd2c1ca0f6a9e46/org.bukkit/plugin/messaging/Messenger.java#L21).

File: src/main/java/fr/xephi/authme/process/changepassword/AsyncChangePassword.java [hunk 2]
```diff
@@ -84,7 +87,9 @@ public void changePasswordAsAdmin(CommandSender sender, final String playerName,
 
         HashedPassword hashedPassword = passwordSecurity.computeHash(newPassword, lowerCaseName);
         if (dataSource.updatePassword(lowerCaseName, hashedPassword)) {
-            bungeeSender.sendAuthMeBungeecordMessage(MessageType.REFRESH_PASSWORD, lowerCaseName);
+            if (dataSource instanceof CacheDataSource) {
+                bungeeSender.sendAuthMeBungeecordMessage(MessageType.REFRESH_PASSWORD, lowerCaseName);
+            }
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/process/changepassword/AsyncChangePassword.java]: 1) Maybe it‘s more elegant to have the bungee sender take care of checking that if statement. The method could be named accordingly. 

2) Can we introduce a method on DataSource for this? Something like `boolean isCached()` or `boolean hasCache()`

File: src/main/java/fr/xephi/authme/service/bungeecord/MessageType.java [hunk 3]
```diff
@@ -1,19 +1,51 @@
...
+
+    /**
+     * Returns the MessageType with the given ID.
+     *
+     * @param id the message type id.
+     * @return the MessageType with the given id, MessageType.UNKNOWN if invalid.
+     */
+    public static MessageType fromId(String id) {
+        for(MessageType current : values()) {
+            if(current.getId().equals(id)) {
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/service/bungeecord/MessageType.java]: Minor: Please add space before reserved keywords like `if` and `for`

File: src/main/java/fr/xephi/authme/service/bungeecord/MessageType.java [hunk 4]
```diff
@@ -1,19 +1,51 @@
...
+                return current;
+            }
+        }
+        return UNKNOWN;
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/service/bungeecord/MessageType.java]: Discussion: does it make sense to return an enum entry here that I guess only exists for this method? Other alternatives could be to return `null` here, or an Optional

Comment (sgdc3) [reply src/main/java/fr/xephi/authme/service/bungeecord/MessageType.java]: not a big fun of nulls, and an optional is a little overkill in this case, idk

File: src/main/java/fr/xephi/authme/service/bungeecord/MessageType.java (hunk 3 above)
Comment (ljacqu) [reply src/main/java/fr/xephi/authme/service/bungeecord/MessageType.java]: Good point; I think there‘s a Checkstyle rule that can be added. I can check once I‘m back home :) (still on vacation)

File: src/main/java/fr/xephi/authme/service/bungeecord/MessageType.java (hunk 4 above)
Comment (ljacqu) [reply src/main/java/fr/xephi/authme/service/bungeecord/MessageType.java]: I saw you changed it – thank you – the reason it irked me is that it‘s an additional entry in the enum but it‘s not really a message type; it‘s a marker for an invalid state, so to speak.

**Comment by sgdc3** [comment]
@ljacqu, any idea why the jdk8 circle build is failing? :/

**Comment by ljacqu** [comment]
The internet here in Kyrgyzstan is so bad I can‘t even load the CircleCI log 😂

**Pull request merged**